### PR TITLE
ceph_mon: "mon host" config should be honored

### PR DESCRIPTION
**Pull request by ErwanAliasr1**
When no mon section are declared, the current behavior was to bind on the public network.
The "mon host" configuration should be honored in such case.
This patch is about reading the "mon host" configuration option before using the public network as fall back.

In such case, a typical output looks like :

	2016-11-25 17:16:17.869587 7f87502c5580  0 mon.ceph-mon0 does not exist in monmap, will attempt to join an existing cluster
	2016-11-25 17:16:17.869621 7f87502c5580  0 using mon host 172.16.0.10:0/0 -> 172.16.0.10:6789/0
	2016-11-25 17:16:17.869723 7f87502c5580  0 starting mon.ceph-mon0 rank -1 at 172.16.0.10:6789/0 mon_data /var/lib/ceph/mon/ceph-ceph-mon0 fsid 8a5bd9e5-001a-431a-9f84-36fef7fc7bc4

fixes: #18043

Tested-by: Sebastien Han <seb@redhat.com>
Signed-off-by: Erwan Velu <erwan@redhat.com>

### CODE REVIEW

File: src/ceph_mon.cc [hunk 1]
```diff
@@ -614,28 +614,42 @@ int main(int argc, const char **argv)
...
-      if (err < 0) {
-	derr << argv[0] << ": error generating initial monmap: "
-             << cpp_strerror(err) << dendl;
-	usage();
-	prefork.exit(1);
+    entity_addr_t conf_host;
+    std::vector <std::string> my_sections;
+    g_conf->get_my_sections(my_sections);
+    std::string mon_host_str;
+    if (g_conf->get_val_from_conf_file(my_sections, "mon host",
```
Comment (leseb) [inline src/ceph_mon.cc]: the patch doesn't address that yet, we will work on that soon.
Thanks!

Comment (ErwanAliasr1) [inline src/ceph_mon.cc]: Good catch. I updated the patch below to parse the mon host and use an interator over it.

Comment (ErwanAliasr1) [inline src/ceph_mon.cc]: My approach was totally wrong. Reworking the patch.

Comment (ErwanAliasr1) [inline src/ceph_mon.cc]: Just update the code with a new approach. As per the commit message says, it does add all the possible IPs to try into a list, add the public network to it and then try to bind against all IPs in the list before making a fault.

**Comment by leseb** [comment]
The last version of that patch has been tested and validated.

**Comment by ErwanAliasr1** [comment]
Hey @jecluis, I just rebased my patch against current master.
I also integrated your remark about the 80 columns as much as possible to keep a readable log.
I hope that is ok for you and +1 for the merge to solve this issue. Cheers,

**Comment by jecluis** [comment]
@ErwanAliasr1 It seems to me that you are trying to reimplement what is already being done in `MonMap::build_initial()` (and `MonMap::build_from_host_list()`), but without honoring `mon_initial_members`.

If I understand it correctly, you are trying to have the monitor bind to any address on `mon_host` if the monitor is not present in the monmap, and trying to then fall back to `public_addr`/`public_network` if all else fails - including generating the initial monmap. This patch does have a few problems though:

1. it seems there's a presumption that the monmap will not include this monitor; this should only happen if a monmap if either fed to mkfs, or if the conf file does not contain the monitor at mkfs-time (either in the form of a mon-specific section, or on `mon_host`/`mon_initial_members`.

2. assuming the monitor is not in the monmap, making your own `mon_host` discovery before anything else will prevent us from entering the build_initial branch if `mon_host` has been provided - thus breaking the expected monmap generation. There are other ways to finding a monitor and populating the monmap, like, as mentioned, pairing `mon_host` and `mon_initial_members`, or figuring out from possibly available sections, and even from dns srv records.

3. The patch is pushing picking an address via `public_addr`/`public_network` to a later time. Whether this should be done is debatable, as it's expected that the monitors give precedence to public_addr (unless specified in the monmap before hand). But the way you're doing it is causing unit tests to fail, given you are not considering the scenario in which only 'public_addr' is specified, instead of mon_host or anything else, and we end up erroring out saying there's `no public_addr or public_network specified`.

TL;DR: I'm not fully grasping what's your motivation to change this behavior, and it feels a bit like it may be because you are not specifying 'mon_initial_members' along with your 'mon_host' on your config file. Either that, or there's some weird bug somewhere, but this somehow doesn't look the right way to fix it - or maybe I'm misunderstanding what you are trying to achieve.

**Comment by ErwanAliasr1** [comment]
@jecluis I'm confused. On one hand I have http://tracker.ceph.com/issues/18043 saying the mon host is not honored leading to a binding to the wrong ip, on the other hand you tell me that it should be honored by the existing code.

I get that my patch seems to go on the wrong way since I'm redoing stuff which is supposed to be part of build_initial(). So the real bug is about build_initial() not taking care properly of "mon host" ?

**Comment by jecluis** [comment]
@ErwanAliasr1 Yeah, both things happen to be correct. I just managed to reproduce this, so let me get back to you later today with some more thoughts.

**Comment by dachary** [comment]
https://jenkins.ceph.com/job/ceph-pull-requests/15230/

<pre>
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:355: run_mon:  ceph-mon --id b --mon-osd-full-ratio=.99 --mon-data-avail-crit=1 --paxos-propose-interval=0.1 --osd-crush-chooseleaf-type=0 --erasure-code-dir=/home/jenkins-build/build/workspace/ceph-pull-requests/build/lib --plugin-dir=/home/jenkins-build/build/workspace/ceph-pull-requests/build/lib --debug-mon 20 --debug-ms 20 --debug-paxos 20 --chdir= --mon-data=td/misc/b '--log-file=td/misc/$name.log' '--admin-socket=td/misc/$cluster-$name.asok' --mon-cluster-log-file=td/misc/log --run-dir=td/misc '--pid-file=td/misc/$name.pid' --mon-allow-pool-delete --public-addr 127.0.0.1:7118
2016-12-01 11:13:51.756416 7f65f2c7f780 -1 no public_addr or public_network specified, and mon.b not present in monmap or ceph.conf
/home/jenkins-build/build/workspace/ceph-pull-requests/qa/workunits/ceph-helpers.sh:373: run_mon:  return 1
/home/jenkins-build/build/workspace/ceph-pull-requests/src/test/mon/misc.sh:119: TEST_mon_add_to_single_mon:  return 1
</pre>
jenkins test this please to assert it is not transient

**Comment by jecluis** [comment]
@dachary it's not transient. The current patch breaks the way 'public_addr' is handled, hence the failure.

**Comment by ErwanAliasr1** [comment]
@jecluis I've been rereading your comments and the code.

In @leseb's case, we never pass through the build_initial as he have a public network defined which explain why he have the bug.

My proposal isn't perfect but I'd like to promote an idea I had to solve the priorities between various configuration options.

I tried to add into a list, all the IP we could bind against and sort them by order of priority. The most priority option coming first in the list and the least priority at the bottom. Then I'm trying to bind against all that IPs. This way it's pretty easy to add new config options without having to handle a new if() else() statement and some options are not always exclusive.

WDYT about this approach ?

**Comment by ErwanAliasr1** [comment]
@jecluis Any update on that topic ?

**Comment by ErwanAliasr1** [comment]
@jecluis ping

**Comment by liewegas** [comment]
@jecluis ping?

**Comment by ErwanAliasr1** [comment]
its sad to see this PR remained in that state, we never got answers or proposal from @jecluis 

What should we do about it ?
does it still have interest to keep it ?
shall we rework it ?

**Comment by ErwanAliasr1** [comment]
@leseb ? Shall we close it then ?

**Comment by liewegas** [comment]
@jecluis ping?

**Comment by leseb** [comment]
@ErwanAliasr1 I can't remember all the details but there is something to fix here, so I'd like to keep it open.

**Comment by jecluis** [comment]
I totally forgot this existed, sorry. At the time, I did spend some time trying to figure out the best way to do this (see https://github.com/jecluis/ceph/commits/wip-mon-host), but for some reason that was put on the back burner and was eventually forgotten. Let me go through this PR later today and check out what it was that I was attempting to do, and see if we can figure out a way to push this forward.

**Comment by ErwanAliasr1** [comment]
@jecluis You totally forgot this PR for 2 years. You told my patch wasn't perfect but you never proposed anything else. Today, this PR surely doesn't have any meaning anymore.
Thanks for the fish.
